open_spiel/python/games/domino.py

```diff
@@ -0,0 +1,346 @@
+    utility_sum=0.0)
+
+
+class Action:
+    """ represent player possible action """
+
+    def __init__(self, tile_to_put, pip_to_play_on, player, edges):
+        self.tile_to_put = tile_to_put
+        self.open_pip = pip_to_play_on
+        self.player = player
+        self.edges = edges
+        self.new_edges = self.edges_after_action()
+
+    def edges_after_action(self):
+        new_edges = []
+        if len(self.edges) == 0:  # first tile on board
+            new_edges.append(self.tile_to_put[0])
+            new_edges.append(self.tile_to_put[1])
+        else:
+            edge_to_stay = self.edges[0] if self.edges[0] != self.open_pip else self.edges[1]
+            new_edge = self.tile_to_put[0] if self.tile_to_put[0] != self.open_pip else self.tile_to_put[1]
+            new_edges.append(edge_to_stay)
+            new_edges.append(new_edge)
+
+        new_edges.sort()
+        return new_edges
+
+    def __str__(self):
+        return f'p{self.player} tile:{self.tile_to_put} pip:{self.open_pip} new_edges:{self.new_edges}'
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class DominoGame(pyspiel.Game):
+    """A Python version of Domino Block."""
+
+    def __init__(self, params=None):
+        super().__init__(_GAME_TYPE, _GAME_INFO, params or dict())
+
+    def new_initial_state(self):
+        """Returns a state corresponding to the start of a game."""
+        return DominoState(self)
+
+    def make_py_observer(self, iig_obs_type=None, params=None):
+        """Returns an object used for observing game state."""
+        return DominoObserver(
+            iig_obs_type or pyspiel.IIGObservationType(perfect_recall=False),
+            params)
+
+
+class DominoState(pyspiel.State):
+    """A python version of the Domino state."""
+
+    def __init__(self, game):
+        """Constructor; should only be called by Game.new_initial_state."""
+        super().__init__(game)
+        self.gameHistory = []
+        self.open_edges = []
+        self.player_legal_actions = []
+        self.hands = [[], []]
+        self.deck = copy.deepcopy(list(_DECK))
+        self._game_over = False
+        self._next_player = pyspiel.PlayerId.CHANCE
+
+    # OpenSpiel (PySpiel) API functions are below. This is the standard set that
+    # should be implemented by every sequential-move game with chance.
+
+    def current_player(self):
+        """Returns id of the next player to move, or TERMINAL if game is over."""
+        if self._game_over:
+            return pyspiel.PlayerId.TERMINAL
+        elif len(self.player_legal_actions) == 0:
+            return pyspiel.PlayerId.CHANCE
+        else:
+            return self._next_player
+
+    def _legal_actions(self, player):
+        """Returns a list of legal actions, sorted in ascending order."""
+        assert player >= 0
+        assert player == self._next_player
+        return list(range(0, len(self.player_legal_actions)))
```

Comment:
I don't see how this can be correct. It always returns a sequential list of actions starting from 0. 

If this is indeed a bug, it highlights the need for more thorough tests because somehow the random sims didn't throw any errors.

Reply:
self.player_legal_actions contains the optional actions for the this state.
the meaning of actions 0 is depend on the actual state.
hope it make sense.

Reply:
Yeah pretty much all games have state-independent interpretations of action ID numbers and that should be the case in all games, for the reasons John says. 

I'm not sure which game it is, can you send me a link? Is it this game? [Block](https://www.dominorules.com/block)

If so, then I counted 27 tiles. These can be given ID's based on their index in that frozen set. Then you can do it two ways. Option 1: the same player moves twice. The first move is picking the tile from hand, sec

Reply:
@lanctot I'm not sure which game it is, can you send me a link? Is it this game? [Block](https://www.dominorules.com/block)
yes this is the game
the only difference is how the player that plays first is chosen (you will find many ways of doing so in different resources).

how about doing something similar to option 1 but instead of indexing the tile we playing and the tile we playing on, indexing the tile we playing and the pip we playing on? 

why is good?
in your idea, we need to know what tile is connected to our open edge to find the index of possible action.

for example, if our board looks like this:
(6,6)(6,5)(5,4)(4,3)

To compute the move index it is not enough to understand that 6 and 3 are the open edges, you also need to know to what tiles these open edges are connected. I like this solution because it's very natural to just track the open edges after each move.

Reply:
By "pip", do you mean it would be 6 or 3 in your example?

What happens if there are two possible places that you can connect the same value pip.. that can happen, right? How do you resolve it? (That is the main reason I suggested the tile id for the connection because they are always unique)

Reply:
Was thinking the same thing. If in the process of resolving the ambiguity you end up tracking something else (like the tiles they're connected to), then you need to also somehow encode this in the action space, the meaning of which is going to be less transparent than the unique tile id.

The interpretability of the action space is usually more important than condensing it to as few actions as possible.

Reply:
@lanctot 
By "pip", do you mean it would be 6 or 3 in your example?
yes.

What happens if there are two possible places that you can connect the same value pip.. that can happen, right? How do you resolve it? (That is the main reason I suggested the tile id for the connection because they are always unique)

here are some examples:

board: (2,3) (3,4) 
tiles in my hand: (2,4) (1,1) (6,6)
possible actions:  
1. (2,4), 2 
board after the move: (4,2) (2,3), (3,4) 
2. (2,4), 4
board after the move:  (2,3) (3,4), (4,2) 

board: (3,3) 
tiles in my hand: (3,4) (1,1) (6,6)
possible actions:  
1. (3, 3), 3  // the end you playing on the tile (right or left) as no meaning strategy-wise so we have single possible action.


That is the main reason I suggested the tile id for the connection because they are always unique
I think both ways can work.
consider the following situation

board after turn 1:  (0,0) 
board after turn 2:  (0,0) (0,1) 
board after turn 3:  (0,0) (0,1) (1,2)
board after turn 4:  (0,0) (0,1) (1,2) (2,3)
board after turn 5:  (6, 0) (0,0) (0,1) (1,2) (2,3)

for me, it feels more natural to think about the last action as: "put (6,0) on 0" instead of "put (0,6) on (0, 0)".

on the other hand, consider spinner variation: the first double of the board has 4 edges. 
so if use my representation. 
in my representation, playing on a spinner will be represented in the same way as playing on a regular tile.

After this, I prefer the 2 tiles representation since I think it will be easier to transfer to other variations.
I will also need to give some Id for the "empty board" (in the first turn)

Reply:
I am afraid that I am not following.

Suppose you have a board like this: 

(3,2) (2,5) (5,5) (5,3)

And you want to play "(3,4) on the 3".

How do you know which 3.. the left one, or right one?

Reply:
I treat it as the same move. strategy-wise these moves are interchangeable.

Reply:
Hmm.. yeah, I see. Ok I'm not super familiar with the game -- are you sure you can't get into a situation where these two moves would not be strategically equivalent?

Reply:
in domino block, I am sure it's strategically equivalent.

the are some (all-5) domino variants that it is not.
but also in all-5 in most cases, these moves will be equivalent.

about all-5: 

in all-5, players score points whenever they play a tile to the board so that the Board Count would then equal any multiple of five. The Board Count (BC) is the sum total of all available open ends on the board. 
(http://www.chessandpoker.com/dominoes-strategy.html).

![image](https://user-images.githubusercontent.com/31830533/223133768-71ded1a4-d204-42bd-9ab1-d8df10f0bf31.png)

in example 3 from the illustration, you can see that the edges of the tile (6,6) are excluded from the board count. the reason for this is that spinners' edges are "burned" when they have two arms connected to them.

now consider the following board:

(1,1) (1,2) (2,3) (3,1)
and your only playable tile is (1,4).
If you will play it on the spinner side (left) the BC will be 5, and you will get points.
If you will play it on the non-spinner side (right) the BC will be 6, and you will not get points.

Reply:
Ok, works for me. (I'm not too concerned about covering all the variants, because if it's extended, we can change the actions later.) @jhtschultz wdyt?

Reply:
Yep works for me too. If you don't have to disambiguate between edges then using that for the action space seems fine and we'll revisit if we implement more variants.